### ISSUE TITLE
Don't allow layer switcher actions on non-base layers

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/agent#216]
On non-base layers, the user must not be allowed to assign layer switcher actions, because they're not only practically useless but outright confusing.

Instead of the current content of the Layer tab of the key action popover, let's display the

>  Layer switching is only possible from the base layer.

text for non-base layers.

**Comment by Deewiant** [comment]
Can I ask to reconsider this?

My use case is that I don't see myself using the Mouse layer commonly enough to justify dedicating a key for it, so I wanted to put it behind Mod, using the Mouse key for something else normally.

I would've imagined it working like so:

* While I'm holding down Mod:
  * While I'm holding down Mouse (or after double-pressing Mouse):
    * Mouse layer is active instead of Mod.
    * When I release Mouse (or after double-pressing it again), the Mod layer is activated.
    * When I release Mod, nothing happens, allowing using this key as intended in the Mouse layer.
* After double-pressing Mod:
  * After double-pressing Mouse:
    * Mouse layer is active instead of Mod.
    * After double-pressing Mouse again, the base layer is activated — I wouldn't expect the layers to nest like in the holding down case.
  * While I'm holding down Mouse:
    * Mouse layer is active instead of Mod.
    * When I release Mouse, the Mod layer is activated.

**Comment by mondalaci** [comment]
@Deewiant Your exact use case doesn't fit into the established framework, and I see no way of implementing it without introducing additional complexity at the expense of everybody else. 

Let me suggest two workarounds that result in quite similar functionality:

1. Create two different keymaps: a Default keymap and a Mouse keymap. On the Default keymap make Mod + Caps Lock switch to the Mouse keymap. On the base layer of the Mouse keymap set up mouse keys and bind a switch keymap action to Caps Lock which switches back to the Default keymap.
2. Alternatively, use a single keymap and make the Caps Lock to emit a scancode of your desire, and also set its secondary role to Mouse.

The first vs second way allows you to reach mouse keys via toggling vs holding without sacrificing an extra key. Hope I could help!

**Comment by Deewiant** [comment]
Fair enough, and thanks for the suggestions! I didn't realize that layer switching was an option for secondary roles, that seems to be a simple solution that works well enough.

**Comment by isometric** [comment]
I think this should be reconsidered also. I want to toggle directly between any two layers such as mouse -> fn -> mod -> default saving a key press and some confusing transferring between non-default layers. Alternatively, having more than one key (other toggle keys) to toggle to default from a non-default layer would also help a bit with the cognitive overhead but not save the keypress. Perhaps it might make sense to make it an advanced-mode feature or some such.

**Comment by mondalaci** [comment]
@isometric I'm sorry, but I don't plan to reconsider this. The current way is by far the simplest way, and almost every user is satisfied with it. Advanced modes usually translate to a lot of extra work, not only in the short run but especially on the long run (mainteance). Given our resources, it's not worth it.

**Comment by isometric** [comment]
I really don't think it would be that much work to just bypass some safeguards or have an official WAR like via the json imports but you do you. I can say I was planning on buying a few more sets but this is a dealbreaker.

**Comment by kareltucek** [comment]
@isometric it would be much work, because this thing is bugged in serializer too and the UHK team does not bother to fix it since it is "non-issue" due to the existence of those safeguards. 

I have a counter-proposal and an alternative solution for the problem, just give me some time - I will hopefully write the things down once I get reply for firmware #199.

**Comment by farzher** [comment]
i often lock myself in the mouse layer, and want easier ways to get out so i can quickly start typing again.
it would be nice if i could bind a bunch of mouse layer keys to switch me back to base layer!

**Comment by mondalaci** [comment]
@farzher A dedicated action to leave the mouse layer (or any layer) is not planned. If you have deboucing issues, feel free to shoot us an email, and we'll help you.

**Comment by kareltucek** [comment]
(@farzher, it is possible via macros on my firmware fork.)

**Comment by jlsjonas** [comment]
@mondalaci  I'm also going to vote to reconsider this, due to layer locking.
use-case:
I often lock myself in the mouse layer, where I often need to _just quickly_ use arrow keys or similar; to accomplish this now I have to unlock the mouse layer, hold-switch to the fn layer to then do my arrow actions.

I should just be able to hit "fn" and use the arrow keys, regardless of what is my current locked layer.

To be clear, I do agree that this could be confusing in a hold-method, hence also proposing this specifically for active *locked* layers (so it feels natural; instead of wonky)

**Comment by mondalaci** [comment]
Switching to a non-base layer directly from a toggled layer makes sense to me usability-wise.

For the sake of simplicity, Agent shouldn't be modified, only the firmware. The current "Layer switching is only possible from the base layer." sentence should be replaced with "Layer switching can only be mapped on the base layer."

On the firmware level, if a layer is toggled, the user could directly switch to a non-base layer when pressing a key which is a layer switch action on the base layer and a none action on the actual layer. No recursive layer switching mechanism should be used for the sake of simplicity.

@kareltucek How much work would it take to implement this change in the firmware?

**Comment by kareltucek** [comment]
> How much work would it take to implement this change in the firmware?

Not sure - hopefully not much. But it still raises red flags for me - acting based on bindings from other layers sounds very wrong. I mean - it's making firmware code more complex for no good reason, why also making the user logic more complex and less intuitive - all at the same time.

What about solving this on Agent level instead? It should suffice to let agent map layer switchers on all layers by default, and then have firmware just do what it is told without having to care about weird-and-complex conditions. Seems much cleaner, and even looks more proficient in preventing new users from getting themselves locked in layers by accident.

On top of it, you may easily decide to allow arbitrary layer switching to advanced users. Or not - your call...

> On the firmware level, if a layer is toggled, the user could directly switch to a non-base layer when pressing a key which is a layer switch action on the base layer and a none action on the actual layer.

So this should work with all types of layer switchers? I.e.:

- held to held
- held to toggled
- toggled to held
- toggled to toggled

Furthermore, let's assume following scenario:

  - Hold (activate) Mod is pressed
  - Hold (activate) Fn is pressed
  - Hold (activate) Fn is released
  - Now, is it vital that keyboard goes back to Mod layer or should it go to Base? Do we care at all? Especially considering your "non-recursivity" requirement, this is becoming a bit interesting question. 
 
 ("Base" or "We don't care" answers might actually allow us to drop the current complex layer hold strategy and simplify things.)

**Comment by jlsjonas** [comment]
> switch action on the base layer and a none action on the actual layer.

That feels potentially confusing configuration-wise, although it would feel intuitive.
alternative approach is to _bleed_ the base layer indication through, but not allow modifying it (to avoid recursiveness-inducing configs)
That should still be fairly simple to implement (UI wise)

>("Base" or "We don't care" answers might actually allow us to drop the current complex layer hold strategy and simplify things.)

I believe that that would defeat the purpose of the suggestion; if I'm locked to the mouse layer, hold the Fn button to use the arrow keys I'd expect to toggle back to the mouse layer when releasing the Fn key

**Comment by kareltucek** [comment]
>> ("Base" or "We don't care" answers might actually allow us to drop the current complex layer hold strategy and simplify things.)
>
> I believe that that would defeat the purpose of the suggestion; if I'm locked to the mouse layer, hold the Fn button to use the arrow keys I'd expect to toggle back to the mouse layer when releasing the Fn key

Actually, that note was on "held to held" logic only (not "toggled to held") - which currently behaves exactly opposite to your proposal -  i.e., in the scenario above, fn layer never gets activated since the active-layer switcher (Mod) has priority for as long as its pressed.

Proposal is to reverse that logic - to allow one hold replace another as soon as it is activated. Yet in my understanding of the issue, held ("activated") layers should always work on top of toggled (locked) layers so that when hold is released, you get returned to your toggled layer. 

Furthermore, if one layer is toggled on top of another toggled one, it should simply replace the layer in the "toggle slot". (This behaviour would probably be valid for advanced configuration only - e.g., via macros - since Laci clearly opposes it.)

**Comment by jlsjonas** [comment]
>so that when hold is released, you get returned to your toggled layer.
>Furthermore, if one layer is toggled on top of another toggled one, it should simply replace the layer in the "toggle slot".

That seems like the most intuitive case indeed.

Some examples, to ensure we understand each other correctly:

in case 1, if the mouse layer is locked (double-tap action), fn button is pressed (held down) => fn becomes active; when fn is released => mouse becomes active again

Case 2 (that would still feel intuitive to me, but might go into terrain that Laci opposes): mouse layer is locked (double-tap action), attempt is made to lock the fn layer (again double-tap): ideally you get locked into the fn layer, which has now replaced the mouse layer (going back to mouse via base, or the mouse layer button if you have it not-assigned/matching base)

Case 1 feels the most important one to solve, though the bridge towards case 2 (to round things up, and likely keep things simpler technically) seems pretty small

**Comment by mondalaci** [comment]
@jlsjonas Now I understand better what you're after. Layer toggles would override the current layer toggle, and held layers would temporarily activate their layers, then recurse to the toggled layer. The suggested behavior is intuitive for you, and it is for me after understanding your use case, but I don't think it'd be intuitive for others due to layer recursion.

In this issue, various use cases have been voiced by multiple users. On second thought, I'd much rather stick to the current behavior and allow more advanced use cases via [smart macros](https://github.com/UltimateHackingKeyboard/firmware/issues/351). Any idea on how to address these needs in such a way, @kareltucek? This is not a priority feature, but it's worth some thinking.

**Comment by kareltucek** [comment]
Sure the fork ( https://github.com/kareltucek/firmware ) allows holding/toggling layers via macros even from non-base layers.

**Comment by kareltucek** [comment]
> but it's worth some thinking.

Or what exactly is the question? 

I guess having a macro action which allows you to hold/toggle layer is the obvious solution.

**Comment by mondalaci** [comment]
@kareltucek I guess providing the exact macros for @jlsjonas's use case would be useful. I'd rather start with the small subset of macro commands mentioned in the smart macros issue. Afterward, we can gradually expand the set of available macro commands, and @jlsjonas would be able to use your fork in the meantime.

**Comment by kareltucek** [comment]
> I guess providing the exact macros for @jlsjonas's use case would be useful.

@mondalaci  I dont think this needs any special macros - just the basic layer switching commands from my fork should do the trick.

Ofc., merging/enabling commands to master is up to you. I am not entirelly sure what to do with the layer switching mechanism yet anyway...
